Apps/GatewayApi/src/Services/DelegateService.cs

```diff
@@ -0,0 +1,95 @@
+        private readonly IDelegateInvitationDelegate delegateInvitationDelegate;
+        private readonly IPatientDetailsService patientDetailsService;
+        private readonly IHashDelegate hashDelegate;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DelegateService"/> class.
+        /// </summary>
+        /// <param name="autoMapper">The injected auto mapper provider.</param>
+        /// <param name="logger">The injected logger.</param>
+        /// <param name="delegateInvitationDelegate">The injected delegate invitation delegate.</param>
+        /// <param name="hashDelegate">The injected hash delegate.</param>
+        /// <param name="patientDetailsService">The injected patient details service.</param>
+        public DelegateService(
+            IMapper autoMapper,
+            ILogger<DelegateService> logger,
+            IDelegateInvitationDelegate delegateInvitationDelegate,
+            IHashDelegate hashDelegate,
+            IPatientDetailsService patientDetailsService)
+        {
+            this.autoMapper = autoMapper;
+            this.logger = logger;
+            this.delegateInvitationDelegate = delegateInvitationDelegate;
+            this.hashDelegate = hashDelegate;
+            this.patientDetailsService = patientDetailsService;
+        }
+
+        /// <inheritdoc/>
+        public async Task<string> CreateDelegateInvitationAsync(string hdid, CreateDelegateInvitationRequest request, CancellationToken ct = default)
+        {
+            ValidationResult validationResult = await new CreateDelegateInvitationRequestValidator().ValidateAsync(request, ct);
+
+            if (!validationResult.IsValid)
+            {
+                throw new ProblemDetailsException(ExceptionUtility.CreateProblemDetails(ErrorMessages.InvalidDelegateInvitationRequest, HttpStatusCode.BadRequest, nameof(DelegateService)));
+            }
+
+            PatientDetails patient = await this.patientDetailsService.GetPatientAsync(hdid, PatientIdentifierType.Hdid, false, ct);
+            string resourceOwnerIdentifier = $"{patient.PreferredName.GivenName} {patient.PreferredName.Surname[0]}";
+            this.logger.LogDebug("Resource owner identifier for delegate invitation: {ResourceOwnerIdentifier}", resourceOwnerIdentifier);
+
+            DelegateInvitation delegateInvitation = this.autoMapper.Map<DelegateInvitation>(request);
+            string sharingCode = VerificationCodeUtility.Generate();
+            IHash hash = this.hashDelegate.Hash(sharingCode);
+            this.logger.LogDebug("Sharing code: {SharingCode} \n Hash: {Hash} \n Salt: {Salt} ", sharingCode, hash?.Hash!, hash?.Salt!);
+
+            delegateInvitation.SharingCodeHash = hash?.Hash!;
+            delegateInvitation.SharingCodeSalt = hash?.Salt!;
+            delegateInvitation.SharingCodeIterations = hash?.Iterations ?? 0;
+            delegateInvitation.SharingCodeHashFunction = hash?.PseudoRandomFunction.ToString()!;
```

Comment:
It seems suspicious that anything here could be null - I'd expect only `hash.Hash` and `hash.Salt` might require the null-forgiving operator based on the nullability of the types in the interface.  (It would require changing the HmacHash method inside HmacHashDelegate to do anything about that.)

But `hash` itself shouldn't be null or need conditional access.
```suggestion
            delegateInvitation.SharingCodeHash = hash.Hash!;
            delegateInvitation.SharingCodeSalt = hash.Salt!;
            delegateInvitation.SharingCodeIterations = hash.Iterations;
            delegateInvitation.SharingCodeHashFunction = hash.PseudoRandomFunction.ToString();
```

Reply:
There is still a Rider warning.

<img width="2521" alt="Screenshot 2023-11-17 at 11 16 07 PM" src="https://github.com/bcgov/healthgateway/assets/58790456/d3193734-d19f-4bd1-98c9-c97b55565be4">